File: startup/dependency_check.py
"""
Startup Module - Dependency and Environment Checks
===================================================
Handles:
- Python package dependency checking and installation
- Tesseract OCR setup (Windows automation)
- Hardware capability detection
- Lazy-loaded - NO side effects on import
- All functions are triggered explicitly by app.py, not on startup

Key functions:
    check_and_install_dependencies() -> bool
    setup_tesseract() -> bool
    get_latest_versions() -> dict
    detect_hardware() -> dict
"""

import subprocess
import sys
import os


def _version_tuple(version_str: str) -> tuple:
    """Convert a version string like '1.2.3' to a tuple of ints for proper comparison."""
    try:
        return tuple(int(x) for x in version_str.split("."))
    except (ValueError, AttributeError):
        return (0,)


# =============================================================================
# DEPENDENCY METADATA
# =============================================================================

REQUIRED_PACKAGES = {
    # Package name: (import_name, min_version, description)
    "streamlit": ("streamlit", "1.28.0", "UI Framework"),
    "requests": ("requests", "2.31.0", "HTTP Client"),
    "psutil": ("psutil", "5.9.0", "Hardware Detection"),
    "pymupdf": ("fitz", "1.23.0", "PDF Extraction"),
    "pytesseract": ("pytesseract", "0.3.10", "OCR Engine"),
    "opencv-python": ("cv2", "4.8.0", "Image Processing"),
    "Pillow": ("PIL", "10.0.0", "Image Library"),
    "numpy": ("numpy", "1.24.0", "Numerical Computing"),
    "sentence-transformers": ("sentence_transformers", "2.2.0", "Embeddings"),
    "beautifulsoup4": ("bs4", "4.12.0", "HTML Parsing"),
}

OPTIONAL_PACKAGES = {
    "qdrant-client": ("qdrant_client", "1.7.0", "Vector Database"),
    "faiss-cpu": ("faiss", "1.7.4", "Vector Search"),
    "faster-whisper": ("faster_whisper", "0.10.0", "Speech Recognition"),
}

# =============================================================================
# TESSERACT OCR CONFIGURATION
# =============================================================================

TESSERACT_VERSION = "5.5.0.20241111"
TESSERACT_URL = f"https://github.com/tesseract-ocr/tesseract/releases/download/5.5.0/tesseract-ocr-w64-setup-{TESSERACT_VERSION}.exe"
TESSERACT_PATHS = [
    r"C:\Program Files\Tesseract-OCR\tesseract.exe",
    r"C:\Program Files (x86)\Tesseract-OCR\tesseract.exe",
]


# =============================================================================
# TESSERACT FUNCTIONS
# =============================================================================


def find_tesseract() -> str:
    """Find Tesseract executable.

    Returns:
        Path to tesseract.exe if found, None otherwise
    """
    import shutil

    # Check if in PATH
    tesseract_in_path = shutil.which("tesseract")
    if tesseract_in_path:
        return tesseract_in_path

    # Check common install locations
    for path in TESSERACT_PATHS:
        if os.path.exists(path):
            return path

    return None


def install_tesseract() -> bool:
    """Download and install Tesseract OCR. Requires admin for silent install.

    Returns:
        True if installation succeeded, False otherwise
    """
    import tempfile
    import urllib.request

    print(f"\n🔧 Tesseract OCR not found. Installing v{TESSERACT_VERSION}...")
    print("   (This is a ~75MB download, please wait...)")

    try:
        installer_path = os.path.join(tempfile.gettempdir(), "tesseract-installer.exe")
        print("  → Downloading from GitHub...")
        urllib.request.urlretrieve(TESSERACT_URL, installer_path)
        print("  → Download complete")

        # Try silent install first (requires admin)
        print("  → Attempting silent install...")
        result = subprocess.run(
            [installer_path, "/S"],
            capture_output=True,
            timeout=300,
            shell=False,
        )

        if result.returncode == 0:
            print("  ✓ Tesseract installed successfully")
            try:
                os.remove(installer_path)
            except Exception as exc:
                _ = exc  # suppressed intentionally
            return True
        else:
            # Silent install failed (likely no admin rights)
            # Launch interactive installer instead
            print("  ⚠ Silent install requires admin. Launching installer...")
            subprocess.Popen([installer_path])
            print("  → Please complete the installation manually.")
            print("  → Default path: C:\\Program Files\\Tesseract-OCR")
            print("  → Restart the app after installation.")
            return False

    except Exception as e:
        print(f"  ✗ Installation error: {e}")
        print("  → Download manually: https://github.com/tesseract-ocr/tesseract/releases")
        return False


def setup_tesseract() -> bool:
    """Find or install Tesseract and configure pytesseract.

    Returns:
        True if Tesseract is available and configured, False otherwise
    """
    tesseract_path = find_tesseract()

    if not tesseract_path:
        # Try to install on Windows
        if sys.platform == "win32":
            if install_tesseract():
                tesseract_path = find_tesseract()
        else:
            print("⚠️ Tesseract not found. Install with: sudo apt install tesseract-ocr")

    if tesseract_path:
        try:
            import pytesseract

            pytesseract.pytesseract.tesseract_cmd = tesseract_path
            print(f"✓ Tesseract configured: {tesseract_path}")
            return True
        except ImportError as exc:
            _ = exc  # suppressed intentionally

    return False


# =============================================================================
# PACKAGE MANAGEMENT FUNCTIONS
# =============================================================================


def get_installed_version(package_name: str) -> str:
    """Get installed version of a package.

    Args:
        package_name: Name of the package (as shown by pip)

    Returns:
        Version string, or None if not found
    """
    try:
        import importlib.metadata

        return importlib.metadata.version(package_name)
    except Exception:
        return None


def check_import(import_name: str) -> bool:
    """Check if a package can be imported.

    Args:
        import_name: Name to use in __import__()

    Returns:
        True if importable, False otherwise
    """
    try:
        __import__(import_name)
        return True
    except ImportError:
        return False


def install_package(package: str, upgrade: bool = False) -> bool:
    """Install a package using pip.

    Args:
        package: Package name or requirement spec (e.g., "numpy>=1.20")
        upgrade: If True, upgrade existing package

    Returns:
        True if successful, False otherwise
    """
    try:
        cmd = [sys.executable, "-m", "pip", "install"]
        if upgrade:
            cmd.append("--upgrade")
        cmd.append(package)
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120, shell=False)
        return result.returncode == 0
    except Exception:
        print(f"Failed to install {package}: {e}")
        return False


def check_and_install_dependencies(show_status: bool = True) -> bool:
    """Check all dependencies and install missing ones.

    Args:
        show_status: If True, print status messages

    Returns:
        True if all required packages are available, False otherwise
    """
    missing = []
    outdated = []
    installed = []

    for pkg_name, (import_name, min_version, desc) in REQUIRED_PACKAGES.items():
        installed_ver = get_installed_version(pkg_name)
        can_import = check_import(import_name)

        if not can_import or installed_ver is None:
            missing.append((pkg_name, min_version, desc))
        elif _version_tuple(installed_ver) < _version_tuple(min_version):
            outdated.append((pkg_name, installed_ver, min_version, desc))
        else:
            installed.append((pkg_name, installed_ver, desc))

    # Auto-install missing packages
    if missing:
        print(f"\n🔧 Installing {len(missing)} missing packages...")
        for pkg_name, min_version, desc in missing:
            print(f"  → Installing {pkg_name}>={min_version} ({desc})...")
            if install_package(f"{pkg_name}>={min_version}"):
                print(f"    ✓ {pkg_name} installed")
                pass
            else:
                print(f"    ✗ Failed to install {pkg_name}")
                pass
                # Upgrade outdated packages
                pass
    if outdated:
        print(f"\n📦 Upgrading {len(outdated)} outdated packages...")
        for pkg_name, curr_ver, min_ver, desc in outdated:
            print(f"  → Upgrading {pkg_name} ({curr_ver} → {min_ver}+)...")
            if install_package(f"{pkg_name}>={min_ver}", upgrade=True):
                print(f"    ✓ {pkg_name} upgraded")
                pass
            else:
                print(f"    ✗ Failed to upgrade {pkg_name}")
                pass
    if show_status and (missing or outdated):
        print("\n✅ Dependency check complete. Please restart the app if packages were installed.\n")

    return len(missing) == 0 and len(outdated) == 0


def get_latest_versions() -> dict:
    """Check PyPI for latest versions of our packages.

    Returns:
        Dict mapping package names to latest versions on PyPI
    """
    import requests

    latest = {}
    for pkg_name in REQUIRED_PACKAGES.keys():
        try:
            r = requests.get(f"https://pypi.org/pypi/{pkg_name}/json", timeout=5)
            if r.status_code == 200:
                latest[pkg_name] = r.json()["info"]["version"]
        except Exception as exc:
            _ = exc  # suppressed intentionally
    return latest


__all__ = [
    "check_and_install_dependencies",
    "setup_tesseract",
    "find_tesseract",
    "install_tesseract",
    "get_latest_versions",
    "REQUIRED_PACKAGES",
    "OPTIONAL_PACKAGES",
]

File: rust_output/src/dependency_check.rs
/// Startup Module - Dependency and Environment Checks
/// ===================================================
/// Handles:
/// - Python package dependency checking and installation
/// - Tesseract OCR setup (Windows automation)
/// - Hardware capability detection
/// - Lazy-loaded - NO side effects on import
/// - All functions are triggered explicitly by app::py, not on startup
/// 
/// Key functions:
/// check_and_install_dependencies() -> bool
/// setup_tesseract() -> bool
/// get_latest_versions() -> dict
/// detect_hardware() -> dict

use anyhow::{Result, Context};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

pub static REQUIRED_PACKAGES: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub static OPTIONAL_PACKAGES: std::sync::LazyLock<HashMap<String, serde_json::Value>> = std::sync::LazyLock::new(|| HashMap::new());

pub const TESSERACT_VERSION: &str = "5.5.0.20241111";

pub const TESSERACT_URL: &str = "f'https://github.com/tesseract-ocr/tesseract/releases/download/5.5.0/tesseract-ocr-w64-setup-{TESSERACT_VERSION}.exe";

pub static TESSERACT_PATHS: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

pub static __ALL__: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

/// Convert a version string like '1.2.3' to a tuple of ints for proper comparison.
pub fn _version_tuple(version_str: String) -> Result<tuple> {
    // Convert a version string like '1.2.3' to a tuple of ints for proper comparison.
    // try:
    {
        /* tuple */ (version_str.split(".".to_string()).map(|s| s.to_string()).collect::<Vec<String>>().iter().map(|x| x.to_string().parse::<i64>().unwrap_or(0)).collect::<Vec<_>>())
    }
    // except (ValueError, AttributeError) as _e:
}

/// Find Tesseract executable.
/// 
/// Returns:
/// Path to tesseract.exe if found, None otherwise
pub fn find_tesseract() -> String {
    // Find Tesseract executable.
    // 
    // Returns:
    // Path to tesseract.exe if found, None otherwise
    // TODO: import shutil
    let mut tesseract_in_path = shutil::which("tesseract".to_string());
    if tesseract_in_path {
        tesseract_in_path
    }
    for path in TESSERACT_PATHS.iter() {
        if os::path.exists(path) {
            path
        }
    }
    None
}

/// Download and install Tesseract OCR. Requires admin for silent install.
/// 
/// Returns:
/// true if installation succeeded, false otherwise
pub fn install_tesseract() -> Result<bool> {
    // Download and install Tesseract OCR. Requires admin for silent install.
    // 
    // Returns:
    // true if installation succeeded, false otherwise
    // TODO: import tempfile
    // TODO: import urllib::request
    println!("\n🔧 Tesseract OCR not found. Installing v{}...", TESSERACT_VERSION);
    println!("{}", "   (This is a ~75MB download, please wait...)".to_string());
    // try:
    {
        let mut installer_path = PathBuf::from(std::env::temp_dir()).join("tesseract-installer.exe".to_string());
        println!("{}", "  → Downloading from GitHub...".to_string());
        urllib::request.urlretrieve(TESSERACT_URL, installer_path);
        println!("{}", "  → Download complete".to_string());
        println!("{}", "  → Attempting silent install...".to_string());
        let mut result = std::process::Command::new("sh").arg("-c").arg(vec![installer_path, "/S".to_string().output().unwrap()], /* capture_output= */ true, /* timeout= */ 300, /* shell= */ false);
        if result.returncode == 0 {
            println!("{}", "  ✓ Tesseract installed successfully".to_string());
            // try:
            {
                std::fs::remove_file(installer_path).ok();
            }
            // except Exception as exc:
            true
        } else {
            println!("{}", "  ⚠ Silent install requires admin. Launching installer...".to_string());
            subprocess::Popen(vec![installer_path]);
            println!("{}", "  → Please complete the installation manually.".to_string());
            println!("{}", "  → Default path: C:\\Program Files\\Tesseract-OCR".to_string());
            println!("{}", "  → Restart the app after installation.".to_string());
            false
        }
    }
    // except Exception as e:
}

/// Find or install Tesseract and configure pytesseract.
/// 
/// Returns:
/// true if Tesseract is available and configured, false otherwise
pub fn setup_tesseract() -> Result<bool> {
    // Find or install Tesseract and configure pytesseract.
    // 
    // Returns:
    // true if Tesseract is available and configured, false otherwise
    let mut tesseract_path = find_tesseract();
    if !tesseract_path {
        if sys::platform == "win32".to_string() {
            if install_tesseract() {
                let mut tesseract_path = find_tesseract();
            }
        } else {
            println!("{}", "⚠️ Tesseract not found. Install with: sudo apt install tesseract-ocr".to_string());
        }
    }
    if tesseract_path {
        // try:
        {
            // TODO: import pytesseract
            pytesseract.pytesseract.tesseract_cmd = tesseract_path;
            println!("✓ Tesseract configured: {}", tesseract_path);
            true
        }
        // except ImportError as exc:
    }
    Ok(false)
}

/// Get installed version of a package.
/// 
/// Args:
/// package_name: Name of the package (as shown by pip)
/// 
/// Returns:
/// Version string, or None if not found
pub fn get_installed_version(package_name: String) -> Result<String> {
    // Get installed version of a package.
    // 
    // Args:
    // package_name: Name of the package (as shown by pip)
    // 
    // Returns:
    // Version string, or None if not found
    // try:
    {
        // TODO: import importlib.metadata
        importlib.metadata.version(package_name)
    }
    // except Exception as _e:
}

/// Check if a package can be imported.
/// 
/// Args:
/// import_name: Name to use in __import__()
/// 
/// Returns:
/// true if importable, false otherwise
pub fn check_import(import_name: String) -> Result<bool> {
    // Check if a package can be imported.
    // 
    // Args:
    // import_name: Name to use in __import__()
    // 
    // Returns:
    // true if importable, false otherwise
    // try:
    {
        __import__(import_name);
        true
    }
    // except ImportError as _e:
}

/// Install a package using pip.
/// 
/// Args:
/// package: Package name or requirement spec (e.g., "numpy>=1.20")
/// upgrade: If true, upgrade existing package
/// 
/// Returns:
/// true if successful, false otherwise
pub fn install_package(package: String, upgrade: bool) -> Result<bool> {
    // Install a package using pip.
    // 
    // Args:
    // package: Package name or requirement spec (e.g., "numpy>=1.20")
    // upgrade: If true, upgrade existing package
    // 
    // Returns:
    // true if successful, false otherwise
    // try:
    {
        let mut cmd = vec![sys::executable, "-m".to_string(), "pip".to_string(), "install".to_string()];
        if upgrade {
            cmd.push("--upgrade".to_string());
        }
        cmd.push(package);
        let mut result = std::process::Command::new("sh").arg("-c").arg(cmd, /* capture_output= */ true, /* text= */ true, /* timeout= */ 120, /* shell= */ false).output().unwrap();
        result.returncode == 0
    }
    // except Exception as _e:
}

/// Check all dependencies and install missing ones.
/// 
/// Args:
/// show_status: If true, print status messages
/// 
/// Returns:
/// true if all required packages are available, false otherwise
pub fn check_and_install_dependencies(show_status: bool) -> bool {
    // Check all dependencies and install missing ones.
    // 
    // Args:
    // show_status: If true, print status messages
    // 
    // Returns:
    // true if all required packages are available, false otherwise
    let mut missing = vec![];
    let mut outdated = vec![];
    let mut installed = vec![];
    for (pkg_name, (import_name, min_version, desc)) in REQUIRED_PACKAGES.iter().iter() {
        let mut installed_ver = get_installed_version(pkg_name);
        let mut can_import = check_import(import_name);
        if (!can_import || installed_ver.is_none()) {
            missing.push((pkg_name, min_version, desc));
        } else if _version_tuple(installed_ver) < _version_tuple(min_version) {
            outdated.push((pkg_name, installed_ver, min_version, desc));
        } else {
            installed.push((pkg_name, installed_ver, desc));
        }
    }
    if missing {
        println!("\n🔧 Installing {} missing packages...", missing.len());
        for (pkg_name, min_version, desc) in missing.iter() {
            println!("  → Installing {}>={} ({})...", pkg_name, min_version, desc);
            if install_package(format!("{}>={}", pkg_name, min_version)) {
                println!("    ✓ {} installed", pkg_name);
                // pass
            } else {
                println!("    ✗ Failed to install {}", pkg_name);
                // pass
                // pass
            }
        }
    }
    if outdated {
        println!("\n📦 Upgrading {} outdated packages...", outdated.len());
        for (pkg_name, curr_ver, min_ver, desc) in outdated.iter() {
            println!("  → Upgrading {} ({} → {}+)...", pkg_name, curr_ver, min_ver);
            if install_package(format!("{}>={}", pkg_name, min_ver), /* upgrade= */ true) {
                println!("    ✓ {} upgraded", pkg_name);
                // pass
            } else {
                println!("    ✗ Failed to upgrade {}", pkg_name);
                // pass
            }
        }
    }
    if (show_status && (missing || outdated)) {
        println!("{}", "\n✅ Dependency check complete. Please restart the app if packages were installed.\n".to_string());
    }
    (missing.len() == 0 && outdated.len() == 0)
}

/// Check PyPI for latest versions of our packages.
/// 
/// Returns:
/// Dict mapping package names to latest versions on PyPI
pub fn get_latest_versions() -> Result<HashMap> {
    // Check PyPI for latest versions of our packages.
    // 
    // Returns:
    // Dict mapping package names to latest versions on PyPI
    // TODO: import requests
    let mut latest = HashMap::new();
    for pkg_name in REQUIRED_PACKAGES.keys().iter() {
        // try:
        {
            let mut r = /* reqwest::get( */&format!("https://pypi.org/pypi/{}/json", pkg_name)).cloned().unwrap_or(/* timeout= */ 5);
            if r.status_code == 200 {
                latest[pkg_name] = r.json()["info".to_string()]["version".to_string()];
            }
        }
        // except Exception as exc:
    }
    Ok(latest)
}
